### PR TITLE
Add initial support for converting static mode positions file.

### PR DESCRIPTION
**Pull request by mattdurham**
#### PR Description

This adds the ability to convert from static to flow with some caveats that I tried to capture in the documentation.

#### Which issue(s) this PR fixes

This is one half of resolving #6323

#### Notes to the Reviewer

There will be a config converter change coming to automatically apply this.

#### PR Checklist

- [x] CHANGELOG.md updated
- [x] Documentation added
- [x] Tests updated

### CODE REVIEW

File: docs/sources/flow/reference/components/loki.source.file.md [hunk 1]
```diff
@@ -37,19 +37,30 @@ log entries to the list of receivers passed in `forward_to`.
...
+
+{{< admonition type="note" >}}
+The `legacy_positions_file` argument is used when you are transitioning from legacy. The legacy positions file will be rewritten into the new format.
+This operation will only occur if the new positions file does not exist and the `legacy_positions_file` is valid. Once converted successfully the `legacy_positions_file` will be deleted.
+If you add any labels before `loki.source.file` then the positions file will conversion will not work. The legacy positions file did not have a concept of labels in the positions file,
+so the conversion assumes no labels.
+{{< /admonition >}}
+
+
+
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/loki.source.file.md:63]:
```suggestion
{{< /admonition >}}

```

Remove empty lines

File: internal/component/common/loki/positions/positions.go [hunk 2]
```diff
@@ -96,6 +96,60 @@ type Positions interface {
...
+// LegacyFile is the copied struct for the static mode positions file.
+type LegacyFile struct {
+	Positions map[string]string `yaml:"positions"`
+}
+
+// ConvertLegacyPositionsFile will convert the legacy positions file to the new format if:
+// 1. There is no file at the newpath
+// 2. There is a file at the legacy path and that it is valid yaml
+// If all the above is true then the legacy file will be deleted.
+func ConvertLegacyPositionsFile(legacyPath, newPath string, l log.Logger) {
```
Comment (erikbaranowski) [inline internal/component/common/loki/positions/positions.go:108]: nit: This one ended up a bit lengthy. Consider breaking it up a bit. For example, `readLegacyPositionsFile` could exist that is functionally similar to `readPositionsFile`

File: internal/component/common/loki/positions/positions_test.go [hunk 3]
```diff
@@ -37,6 +39,90 @@ func tempFilename(t *testing.T) string {
...
+	for k, v := range ps {
+		require.True(t, k.Path == "/tmp/random.log")
+		require.True(t, v == "17623")
+	}
+	// Ensure old file is deleted.
+	_, err = os.Stat(legacy)
+	require.True(t, os.IsNotExist(err))
+}
+
+func TestConversionWithNewFile(t *testing.T) {
```
Comment (erikbaranowski) [inline internal/component/common/loki/positions/positions_test.go:70]: something is a little unobvious with the test names especially this one. brainstorming here, what about in order:
- `TestLegacyConversion`
- `TestLegacyConversionNoOp` or `TestLegacyConversionPreexisting`
- `TestLegacyConversionNoFile`

File: internal/component/common/loki/positions/positions_test.go [hunk 4]
```diff
@@ -37,6 +39,90 @@ func tempFilename(t *testing.T) string {
 	return name
 }
 
+func TestConversion(t *testing.T) {
+	tmpDir := t.TempDir()
```
Comment (erikbaranowski) [inline internal/component/common/loki/positions/positions_test.go:43]: nit: There's a good bit of similar code that maybe could be pulled into a common util func.

File: internal/component/loki/source/file/file_test.go [hunk 5]
```diff
@@ -314,3 +325,130 @@ func TestEncoding(t *testing.T) {
 		"expected positions.yml file to be written eventually",
 	)
 }
+
+// TestFullConversion tests using static mode to tail a log, then converting and then writing new entries and ensuring
+// the previously tailed data does not come through.
+func TestFullConversion(t *testing.T) {
```
Comment (erikbaranowski) [inline internal/component/loki/source/file/file_test.go:331]: This a very cool end to end test. The func length is quite long which isn't the end of the world but can we add some additional comments about each section in the lower half of the func. Similar to the first half I think they just dropped off.

File: internal/component/loki/source/file/file_test.go [hunk 6]
```diff
@@ -314,3 +325,130 @@ func TestEncoding(t *testing.T) {
...
+	positionsDir := t.TempDir()
+	tmpFileDir := t.TempDir()
+
+	tmpFile, err := os.CreateTemp(tmpFileDir, "*.log")
+	require.NoError(t, err)
+
+	//
+	// Listen for push requests and pass them through to a channel
+	//
+
```
Comment (erikbaranowski) [inline internal/component/loki/source/file/file_test.go:344]: nit: extra line

**Review by erikbaranowski — COMMENTED** [review]
Looking good, some general nits to check out to finalize it. Are we planning on wiring this field up to the converter code in the next PR? That should be a fairly small amount of code so I'm not too worried about if we are doing now or later.

File: internal/component/loki/source/file/file_test.go (hunk 5 above)
Comment (mattdurham) [reply internal/component/loki/source/file/file_test.go:331]: This one ended up finding a bug so I was very glad I did it, will clean it up some.

File: internal/component/common/loki/positions/positions_test.go (hunk 3 above)
Comment (mattdurham) [reply internal/component/common/loki/positions/positions_test.go:70]: Good call

**Comment by mattdurham** [comment]
Hoping to wire it up to the converter next PR.

**Comment by mattdurham** [comment]
Moved the OS test to linux only since it  tended not to clean up the temp directories, ran the test on my mac no problem.

**Pull request merged**